### DISCUSSION TITLE
How to implement automated testing into jsPsych web app?

### DISCUSSION THREAD

**Discussion by Zio-4** [jspsych/jsPsych#3011] [General]
I want to do some unit/integration and e2e tests on some jsPsych web apps. They are all set up with npm and use Webpack as a build tool and are hosted on Firebase. How would you implement tests in an app setup this way? I know how you would do it with a UI framework but I'm not sure how to do it in this instance. I saw this [article ](https://dev.to/thawkin3/how-to-unit-test-html-and-vanilla-javascript-without-a-ui-framework-4io)but the only HTML file I would have access to is from the build folder. Would I have to do something similar to that?

**Comment by jodeleeuw** [comment]
Hi @Zio-4 ,

We have a lot of tests in the main codebase. The most useful examples are probably in the packages/plugin-* folders, where we run something akin to integration tests for most plugins.

**Reply by jodeleeuw** [reply]
Here's one example: https://github.com/jspsych/jsPsych/blob/main/packages/plugin-html-keyboard-response/src/index.spec.ts
